### PR TITLE
Support export of function declarations

### PR DESCRIPTION
**Pull request by selaux**
Currently it is only two failing tests...

### CODE REVIEW

**Comment by selaux** [comment]
Fixed.

File: test/unit/BabelPluginExportToFunctionSpec.js [hunk 1]
```diff
@@ -72,7 +95,17 @@ test('BabelPluginExportToFunction should inject config ast', (t) => {
...
+
+    t.is(error.message, 'unknown: auth0-bundler is missing default export in module');
+});
+
+test('BabelPluginExportToFunction should throw on variable export', (t) => {
+    const exportToFunction = new BabelPluginExportToFunction(babelTypes.stringLiteral('testconfig'));
+    const code = 'const a = "test"; exports.default = a;';
+    const error = t.throws(() => babel.transform(code, { plugins: [ exportToFunction ] }));
+
+    t.is(error.message, 'unknown: auth0-bundler is missing default export in module');
```
Comment (lo1tuma) [inline test/unit/BabelPluginExportToFunctionSpec.js]: I’m not sure if I understand what the error message tries to explain.
Isn’t the error that `auth0-bundler` is not able to resolve a function that is exported?
If so I would suggest something like this:

`Error: auth0-bundler could not resolve a function that is being exported.`

File: test/unit/BabelPluginExportToFunctionSpec.js [hunk 2]
```diff
@@ -82,7 +115,9 @@ test('BabelPluginExportToFunction should throw on not enough arguments', (t) =>
         'exports.default = rule;'
     ].join('\n');
 
-    t.throws(() => babel.transform(code, { plugins: [ exportToFunction ] }));
+    const error = t.throws(() => babel.transform(code, { plugins: [ exportToFunction ] }));
+
+    t.is(error.message, 'unknown: auth0-bundler is missing default export in module');
```
Comment (lo1tuma) [inline test/unit/BabelPluginExportToFunctionSpec.js]: Why is this still expected to be an error?
I also don’t understand what the error message has to do with "not enough arguments".

File: lib/BabelPluginExportToFunction.js [hunk 3]
```diff
@@ -37,11 +50,14 @@ function BabelPluginExportToFunction(configAst) {
...
                 if (isDefaultExport(left) && isIdentifier(right)) {
                     const functionIdentifier = path.node.right;
-                    const binding = path.scope.bindings[functionIdentifier.name];
-
-                    if (binding && isValidFunctionExpression(binding.path.node.init)) {
-                        const functionDefinition = binding.path.node.init;
+                    const node = recursivelyResolveBinding(path.scope.bindings, functionIdentifier);
+                    const functionDefinition = R.cond([
+                        [ isValidFunctionDeclaration, R.identity ],
+                        [ R.pipe(R.propOr({}, 'init'), isValidFunctionExpression), R.prop('init') ],
```
Comment (lo1tuma) [inline lib/BabelPluginExportToFunction.js]: Maybe `R.propOr({}, 'init')` and `R.prop('init')` can be extracted, as it is used several times.

Comment (lo1tuma) [inline lib/BabelPluginExportToFunction.js]: Maybe out of scope of this PR, but does `isValidFunctionExpression` support arrow functions?

File: lib/BabelPluginExportToFunction.js [hunk 4]
```diff
@@ -18,15 +18,28 @@ const isDefaultExport = R.allPass([
...
     R.pathEq([ 'property', 'name' ], 'default')
 ]);
+const hasMoreThanTwoParameters = R.pathSatisfies(R.gte(R.__, 2), [ 'params', 'length' ]);
 const isValidFunctionExpression = R.allPass([
     R.propEq('type', 'FunctionExpression'),
-    R.pipe(R.prop('params'), R.length, R.gte(R.__, 2))
+    hasMoreThanTwoParameters
+]);
+const isValidFunctionDeclaration = R.allPass([
+    R.propEq('type', 'FunctionDeclaration'),
```
Comment (lo1tuma) [inline lib/BabelPluginExportToFunction.js]: This includes generator functions. Not sure if this could become a problem though.

File: lib/BabelPluginExportToFunction.js [hunk 5]
```diff
@@ -18,15 +18,28 @@ const isDefaultExport = R.allPass([
...
     R.map(buildArgumentIdentifier),
     R.append(CALLBACK_ID)
 )(R.range(0, functionDefinition.params.length - 2));
+const recursivelyResolveBinding = R.curry((bindings, identifier) => {
+    const resolved = bindings[identifier.name];
+    return R.ifElse(
+        R.pipe(R.propOr({}, 'init'), isIdentifier),
+        R.pipe(R.prop('init'), recursivelyResolveBinding(bindings)),
+        R.identity
+    )(resolved.path.node);
```
Comment (lo1tuma) [inline lib/BabelPluginExportToFunction.js]: FWIW: this creates a lot functions during each iteration of the recursion. But I wouldn’t expect a lot of iterations anyway.

File: lib/BabelPluginExportToFunction.js (hunk 3 above)
Comment (selaux) [inline lib/BabelPluginExportToFunction.js]: Currently not. I would do this in another PR if necessary

**Comment by selaux** [comment]
@lo1tuma PTAL.
